src/runtime_src/core/tools/xbtracer/src/lib/capture.cpp

```diff
@@ -0,0 +1,617 @@
+  auto load_func_addr() -> int;
+  auto load_symbols() -> int;
+
+  static std::shared_ptr<router> get_instance()
+  {
+    static auto ptr = std::make_shared<router>();
+    return ptr;
+  }
+
+  router()
+  : path("")
+  {
+    if (!load_symbols())
+    {
+      std::cout << "Failed to load symbols exiting application " << std::endl;
+      throw std::runtime_error("Failed to load symbols");
+    }
+    else if (!load_func_addr())
+    {
+      std::cout << "Failed to load function address exiting application "
+                << std::endl;
+      throw std::runtime_error("Failed to load function address");
+    }
+  }
+
+  ~router()
+  {
+    if (handle)
+      dlclose(handle);
+  }
+};
+
+/* Warning: initialization of 'dptr' with static storage duration may throw
+ * an exception that cannot be caught
+ * The singleton pattern implemented with a static local variable in
+ * get_instance ensures safe, one-time initialization.
+ */
+// NOLINTNEXTLINE(cert-err58-cpp)
+const auto dptr = router::get_instance();
+
+/**
+ * This function demangles the input mangled function.
+ */
+static std::string demangle(const char* mangled_name)
+{
+  int status = 0;
+  char* demangled_name =
+      abi::__cxa_demangle(mangled_name, nullptr, nullptr, &status);
+  if (status == 0)
+  {
+    // Use std::unique_ptr with a custom deleter to manage demangled_name
+    std::unique_ptr<char, decltype(&std::free)> demangled_ptr(demangled_name,
+                                                              std::free);
+
+    /* Conditioning of demangled name because of some platform based deviation
+      */
+    std::vector<std::pair<std::string, std::string>> replacements =
+      {
+        {"std::__cxx11::basic_string<char, std::char_traits<char>, "
+              "std::allocator<char> >", "std::string"},
+        {"[abi:cxx11]", ""},
+        {"std::map<std::string, unsigned int, std::less<std::string >, "
+            "std::allocator<std::pair<std::string const, unsigned int> > > "
+            "const&", "xrt::hw_context::cfg_param_type const&"}
+      };
+
+    std::string result =
+        find_and_replace_all(std::string(demangled_name), replacements);
+
+    return result;
+  }
+  else
+    // Demangling failed
+    return {mangled_name};  // Return the original mangled name
+}
+
+static std::string find_library_path()
+{
+  std::lock_guard<std::mutex> lock(env_mutex);
+  // NOLINTNEXTLINE(concurrency-mt-unsafe) - protected by env_mutex
+  char* ld_preload = getenv("LD_PRELOAD");
+  if (ld_preload == nullptr)
+  {
+    std::cout << "LD_PRELOAD is not set." << std::endl;
+    return "";
+  }
+
+  std::string full_path = std::string(ld_preload);
+  full_path.erase(std::remove(full_path.begin(), full_path.end(), ' '),
+                  full_path.end());
+  return full_path;
+}
+
+/**
+ * This function will update the dispatch table
+ * with address of the functions from original
+ * library.
+ */
+int router::load_func_addr()
```

Comment:
My preference on error codes is not to use them and instead throw on error.  Don't know for sure if exceptions fit here?

Reply:
Sure, we have made the changes. Please check the lasted updated (9/26)